File: src/main/scala/org/deeplearning4j/scala/ImageClassifierExample.scala
package org.deeplearning4j.scala

import scala.util.control.Breaks._
import org.canova.image.recordreader.ImageRecordReader
import org.canova.api.split.FileSplit
import java.io.File
import org.deeplearning4j.datasets.canova.RecordReaderDataSetIterator
import java.util
import au.com.bytecode.opencsv.{CSVReadProc, CSV}
import org.deeplearning4j.nn.conf.NeuralNetConfiguration
import org.nd4j.linalg.lossfunctions.LossFunctions
import org.deeplearning4j.nn.api.{Classifier, OptimizationAlgorithm}
import org.deeplearning4j.nn.weights.WeightInit
import org.deeplearning4j.nn.layers.factory.LayerFactories
import org.deeplearning4j.optimize.listeners.ScoreIterationListener
import org.nd4j.linalg.dataset.api.DataSet
import org.deeplearning4j.eval.Evaluation
import scala.util.control.Breaks


object ImageClassifierExample {

  def main(args: Array[String]) {

    // Setting the file path to data
    val recordReader = new ImageRecordReader(56, 56, true)
    val home = System.getProperty("user.home")
    recordReader.initialize(new FileSplit(new File(home, "lfw")))

    // Initializing iterator with recordreader and batch size
    val iterator = new RecordReaderDataSetIterator(recordReader, 200)

    // Setting up the configuration for logistic regression TODO: Change the classifier for images
    val conf = new NeuralNetConfiguration.Builder()
      .lossFunction(LossFunctions.LossFunction.MCXENT).optimizationAlgo(OptimizationAlgorithm.GRADIENT_DESCENT)
      .activationFunction("softmax")
      .iterations(100).weightInit(WeightInit.ZERO)
      .learningRate(1e-1).nIn(4).nOut(3).layer(new org.deeplearning4j.nn.conf.layers.OutputLayer()).build()

    // Creating the layer using the specified configuration
    val classifier = LayerFactories.getFactory(conf).create(conf).asInstanceOf[Classifier]

    // Iterating over the data and training the model
    val testSet = scala.collection.mutable.MutableList[DataSet]()
    var counter = 0
    val loop = new Breaks
    loop.breakable{
      while (iterator.hasNext) {
        counter += 1
        print(counter)
        if(counter == 67){  // TODO: is this a bug? Last batch doesn't work
          loop.break;
        }
        val dataset = iterator.next()
        val trainSize = (dataset.getFeatureMatrix().size(0) * 0.9).asInstanceOf[Int]
        val featureMatrix = dataset.getFeatureMatrix()
        val trainTest = dataset.splitTestAndTrain(trainSize)
        trainTest.getTrain().normalizeZeroMeanZeroUnitVariance()
        classifier.asInstanceOf[Classifier].fit(trainTest.getTrain().getFeatureMatrix())
        testSet += trainTest.getTest()
      }
    }

    // Assessing the model's performance on test data
     val testIterator = testSet.iterator
     val testData = testIterator.next()
     val testFeatureMatrix = testData.getFeatureMatrix()
     testData.normalizeZeroMeanZeroUnitVariance()
     val eval = new Evaluation()
    }


}

File: src/main/java/datapipelines/ImageClassifierExample.java
package datapipelines;

import org.canova.api.records.reader.RecordReader;
import org.canova.api.split.FileSplit;
import org.canova.image.recordreader.ImageRecordReader;
import org.deeplearning4j.datasets.canova.RecordReaderDataSetIterator;
import org.deeplearning4j.eval.Evaluation;
import org.deeplearning4j.nn.api.OptimizationAlgorithm;
import org.deeplearning4j.nn.conf.MultiLayerConfiguration;
import org.deeplearning4j.nn.conf.NeuralNetConfiguration;
import org.deeplearning4j.nn.conf.distribution.NormalDistribution;
import org.deeplearning4j.nn.conf.override.ConfOverride;
import org.deeplearning4j.nn.multilayer.MultiLayerNetwork;
import org.deeplearning4j.nn.weights.WeightInit;
import org.deeplearning4j.optimize.api.IterationListener;
import org.deeplearning4j.optimize.listeners.ScoreIterationListener;
import org.nd4j.linalg.api.ndarray.INDArray;
import org.nd4j.linalg.dataset.DataSet;
import org.nd4j.linalg.dataset.api.iterator.DataSetIterator;
import org.nd4j.linalg.lossfunctions.LossFunctions;
import java.io.File;
import java.io.IOException;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

import org.slf4j.LoggerFactory;


public class ImageClassifierExample {

    public static void main(String[] args) throws IOException, InterruptedException {


        // Path to the labeled images
        String labeledPath = System.getProperty("user.home")+"/lfw";
         List<String> labels = new ArrayList<>();
        for(File f : new File(labeledPath).listFiles()) {
            labels.add(f.getName());
        }
        // Instantiating a RecordReader pointing to the data path with the specified
        // height and width for each image.
        RecordReader recordReader = new ImageRecordReader(28, 28, true,labels);
        recordReader.initialize(new FileSplit(new File(labeledPath)));

        // Canova to Dl4j
        DataSetIterator iter = new RecordReaderDataSetIterator(recordReader, 784,labels.size());

        // Creating configuration for the neural net.
        MultiLayerConfiguration conf = new NeuralNetConfiguration.Builder()
                .optimizationAlgo(OptimizationAlgorithm.CONJUGATE_GRADIENT)
                .constrainGradientToUnitNorm(true)
                .weightInit(WeightInit.DISTRIBUTION)
                .dist(new NormalDistribution(1,1e-5))
                .iterations(100).learningRate(1e-3)
                .nIn(784).nOut(labels.size())
                .visibleUnit(org.deeplearning4j.nn.conf.layers.RBM.VisibleUnit.GAUSSIAN)
                .hiddenUnit(org.deeplearning4j.nn.conf.layers.RBM.HiddenUnit.RECTIFIED)
                .layer(new org.deeplearning4j.nn.conf.layers.RBM())
                .list(4).hiddenLayerSizes(600, 250, 100).override(3, new ConfOverride() {
                    @Override
                    public void overrideLayer(int i, NeuralNetConfiguration.Builder builder) {
                        if (i == 3) {
                            builder.layer(new org.deeplearning4j.nn.conf.layers.OutputLayer());
                            builder.activationFunction("softmax");
                            builder.lossFunction(LossFunctions.LossFunction.MCXENT);

                        }
                    }
                }).build();

        MultiLayerNetwork network = new MultiLayerNetwork(conf);
        network.setListeners(Arrays.<IterationListener>asList(new ScoreIterationListener(10)));

        // Training
        while(iter.hasNext()){
            DataSet next = iter.next();
            network.fit(next);
        }

        // Testing -- We're not doing split test and train
        // Using the same training data as test.
        iter.reset();
        Evaluation eval = new Evaluation();
        while(iter.hasNext()){
            DataSet next = iter.next();
            INDArray predict2 = network.output(next.getFeatureMatrix());
            eval.eval(next.getLabels(), predict2);
        }

        System.out.println(eval.stats());
    }
}
